### PR TITLE
Initial support for php 7

### LINKED ISSUE #185: PHP 7-dev Error
This may a bit early, but better sooner than later. When trying to start psysh in PHP7-dev, you see:

```
Fatal error: "Scalar\String" cannot be used as a type declaration in phar:///app/php/bin/psysh/vendor/nikic/php-parser/lib/PhpParser/PrettyPrinter/Standard.php on line 86
```

### PR DESCRIPTION
**Pull request by GrahamCampbell**
Closes #185.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Has anyone got any idea what's going on with the tests here?

**Comment by bobthecow** [comment]
Interesting. Apparently PHP Parser (as of 1.2.0) now rewrites things based on the `use` statements? The changelog mentions a "use builder" and "namespace builder", and I imagine they're the culprits.

The tests should prolly support both ways, since PsySH works (and will continue to work) with older versions of PHP Parser.

**Comment by bobthecow** [comment]
Or maybe it's that PHP-Parser's use statement parsing is no longer compatible, so the UseStatement CodeCleaner pass isn't able to rewrite use statements. I bet that's it.

**Comment by GrahamCampbell** [comment]
@bobthecow Though, that means our current release is broken though - we're using 1.x, so we're getting all changes from 1.2.x.

**Comment by bobthecow** [comment]
In other words, 1.2 is a BC break? Sweet.

**Comment by GrahamCampbell** [comment]
@bobthecow It's not meant to be? Ping @nikic.

**Comment by bobthecow** [comment]
The problem with this is that we have to yank all previous releases of PsySH to keep things from resolving to an older version of PsySH if there's a dependency conflict. I wish people were better at SemVer :-/

**Comment by GrahamCampbell** [comment]
@bobthecow Give @nikic some time. It's possible we can fix this in the next patch release of php parser.

**Comment by nikic** [comment]
The problem is likely the check for `Traversable` in https://github.com/bobthecow/psysh/blob/master/src/Psy/CodeCleaner/UseStatementPass.php#L78 - nodes are no longer `Traversable` (but they can still be traversed, like any other object).

You should be able to safely drop that line - it doesn't make sense anyway, as you already typehint against `Node`.

**Comment by bobthecow** [comment]
@nikic Awesome. Tests pass when that check is replaced with an `else`.

**Comment by GrahamCampbell** [comment]
I'll send a pull fix that separately then.

**Comment by bobthecow** [comment]
:+1: Awesome. Thanks all. @nikic, thanks for the quick reply.

**Comment by GrahamCampbell** [comment]
@bobthecow https://github.com/bobthecow/psysh/pull/188

**Comment by GrahamCampbell** [comment]
Also, that php7 thing needs a new minor release rather than a patch release since we're renaming classes.

**Comment by GrahamCampbell** [comment]
Alrightly. How does this look now?

File: composer.json [hunk 1]
```diff
@@ -14,8 +14,8 @@
     ],
     "require": {
         "php": ">=5.3.0",
-        "symfony/console": "~2.3.10|~2.4.2|~2.5",
-        "nikic/php-parser": "~1.0",
+        "symfony/console": "~2.3.10|^2.4.2",
+        "nikic/php-parser": "^1.2.1",
```
Comment (bobthecow) [inline composer.json]: How long ago did the carat operator ship? I checked the composer changelog but didn't see it.

Comment (GrahamCampbell) [inline composer.json]: Ages ago now. Probably over 2 months. Let me check for you...

**Comment by bobthecow** [comment]
You know, if we're going to do a minor version bump, I'd like to see if we can get some of the other PRs into it. Maybe we ship a point release with the PHP-Parser fix first, then get this out after?

**Comment by GrahamCampbell** [comment]
@bobthecow Fine with me. :)

**Comment by bobthecow** [comment]
:+1: this PR looks good though.

File: composer.json (hunk 1 above)
Comment (GrahamCampbell) [inline composer.json]: It was 8th December 2014: https://github.com/composer/composer/pull/2596.

Comment (bobthecow) [inline composer.json]: :+1: Sweet. Works for me.

**Comment by bobthecow** [comment]
v0.4.4 is out with the initial fix, I'll merge this one into develop and we'll get started on v0.5.0 changes :)

**Pull request merged**